torch/csrc/api/include/torch/data/samplers/custom_index.h

```diff
@@ -0,0 +1,17 @@
+#pragma once
+
+#include <cstddef>
+
+namespace torch {
+namespace data {
+namespace samplers {
+/// A base class for custom index types.
+struct CustomIndex {
+  virtual ~CustomIndex() = default;
+
+  /// The number of elements accessed by this index.
+  virtual size_t size() const = 0;
```

Comment:
To be honest this is a very confusing API. Index is usually something that refers to a single entity, not a number of them. Again, having a `Batch` part somewhere would clarify that it's some kind of a list-like thing.